[ACM-7126]: Node labeling guidelines (#5445)
open-cluster-management/rhacm-docs: clusters/hosted_control_planes/hosted-cluster-workload-distributing.adoc

[#hosted-cluster-workload-distributing]
= Distributing hosted cluster workloads

> Before you get started with hosted control planes for {ocp-short}, you must properly label nodes so that the pods of hosted clusters can be scheduled into infrastructure nodes. Node labeling is also important for the following reasons:

Reviewer: @zanetworker why do we want pods of hosted clusters to be scheduled into infrastructure nodes? This seems confusing to me. They're not infrastructure nodes from the management cluster's perspective for sure.

Reviewer 2: To avoid charing them towards cores of a subscription.

* To ensure high availability and proper workload deployment

> * To ensure that control planes are well contained

Reviewer: I would change the above sentence to something like:
* To ensure that control plane workloads have the proper isolation level

This requires some explanation as well. By isolation level I mean one of:
1. Everything shared (control planes for hosted clusters can run on any node designated for control planes.
2. Request serving isolation (request serving pods in their own dedicated nodes)
3. Nothing shared (every control plane in its own dedicated nodes)

I would also add:
* To ensure that control plane workloads are isolated from other workloads in the management cluster.

Reviewer 2: I think "isolation" might be mis-interpreted here. OpenShift already provides necessary means for isolation via SELinux, Network Policies,... which the control planes of hosted clusters are consumers of being a workload. To avoid confusion, I would opt-out of mentioning "contained" or "isolation" except for special cases where it makes sense for the Hosted cluster tenants.

Reviewer 3: this all makes sense. I'd have clear separated points one for
"To ensure that control plane workloads are isolated from other workloads in the management cluster." via `hypershift.openshift.io/hypershift.openshift.io/control-plane: true`
an another one for the different multi-tenancy workload distribution modes Cesar mentions above.

[#hosted-cluster-labels-taints-overview]
== Labels and taints for management cluster nodes

> //lahinson - sept. 2023 - do we need to add a note to prevent customers from using the management cluster for their workload?

Reviewer 2: Yes please :)

Reviewer: I would temper that by saying that they could run other workloads as long as those workloads do not run on nodes where control planes run (which is why it is important to taint nodes for control planes)

Reviewer 2: @csrwng what about the impact on the API-server, etcd of the management cluster?

Reviewer: There is definitely an impact, but that is true whether you only run control plane workloads or other workloads as well.. it's more of what it is that you're running. You need to properly size your MC's control plane for whatever it is that you want to run.

Reviewer 2: @csrwng except that the control-planes are relatively vetted (and perf/scaled) workloads that are more predictible than wild-card workloads a customer might decide to run on the cluster.

Reviewer 2: Knowing how the CP behaves understress, configuring it with the right reqs/limits (which we do), providing the right quotas,... is something we can do for workloads we are familiar with and have properly tested, unlike just any other workloads which might not be properly configured and yields more surprises.

> ** Nothing shared: Every control plane has its own dedicated nodes. 

Reviewer 2: Request serving/nothing is shared can be confusing for self-managed. I would not add them before we have made sure we have that as a supported configuration in self-managed.

Author: Commented out for now.

As a management cluster administrator, you can use the following labels and taints in the management cluster nodes to schedule the control plane workload:

//lahinson - sept. 2023 - If we want to explain the implications of each label and taint, I'll need that info.

> * `hypershift.openshift.io/hypershift.openshift.io/control-plane: true`

Reviewer: s / hypershift.openshift.io/hypershift.openshift.io/control-plane: true / hypershift.openshift.io/control-plane: true /

Reviewer: For this one, it would be good to explain that you need *both* the label and the taint. The effect of this one is that it dedicates a node to running hosted control plane workloads only. This is important if you don't want other workloads in the management cluster (including MCE/ACM, monitoring, etc) to run in your control plane nodes. I would almost say it's required.

> * `node-role.kubernetes.io/infra`

Reviewer: Not sure why this one is here. @zanetworker can you remind me why we wanted control plane workloads in infra nodes?

Reviewer 2: Not to count towards the subscription (additional cores).

Reviewer: For this one we only need the label

> * `topology.kubernetes.io/zone`

Reviewer: This one is only the label that is needed. It is a required label if you are running control planes with a controllerAvailabilityPolicy of HighlyAvailable. For environments where zones don't make sense, you can set this label to the host name of the node. If this label is not present, HighlyAvailable control planes will not deploy successfully.

> * `hypershift.openshift.io/cluster: ${HostedControlPlane Namespace}`

Reviewer 2: @enxebre can you help @lahinson here explain the implications of each label and taint?

Reviewer: This one is both a label and a taint. It should only be used when you want to dedicate a node to a single HostedCluster.

> * `hypershift.openshift.io/control-plane: true`: Use this label and taint to dedicate a node to running hosted control plane workloads. Setting this label and taint is required if you do not want other workloads on the management cluster to run in your control plane nodes.

Reviewer 2: We should then try to articulate that we need to label the nodes that will host the control-plane pods by two things: 

- `hypershift.openshift.io/control-plane`: true to avoid sharing those nodes with others components (e.g., the management cluster's infra components OR any other mistakenly deployed workload).
- `node-role.kubernetes.io/infra`

Author: I made some updates to the text on lines 18 - 26. PTAL.

> * `node-role.kubernetes.io/infra`: Use this label to avoid having a node count toward your subscription.

Reviewer 2: I think we can try to call this out earlier.
```suggestion
* `node-role.kubernetes.io/infra`: Use this label to avoid having the control-plane workload count toward your subscription.
```

Author: I added a mention of this to the first bullet point in the list that starts on line 6.

> * `topology.kubernetes.io/zone`: Use this label to deploy highly available clusters. The zone might be a location, rack name, or the hostname of the node where the zone is set.

Reviewer 2: Makes sense. 

It would be great if we can summon an example and add it here. @enxebre @csrwng can you help with that?

Author: Agreed!

Reviewer: Would this be sufficient?

Given a management cluster with nodes: `worker-1a`, `worker-1b`, `worker-2a`, `worker-2b` where `worker1-a` and `worker-1b` live in `rack1`; and `worker-2a` and `worker-2b` live in `rack2`, use each rack as an availability zone:

```
oc label node/worker-1a node/worker-1b topology.kubernetes.io/zone=rack1
oc label node/worker-2a node/worker-2b topology.kubernetes.io/zone=rack2
```

Author: Works for me. Thank you!

Author: I added Cesar's example.

Pods for a hosted cluster have tolerations, and the scheduler uses affinity rules to schedule them. Pods tolerate taints for `control-plane` and the `cluster` for the pods. The scheduler prioritizes the scheduling of pods into nodes that are labeled with `hypershift.openshift.io/control-plane` and `hypershift.openshift.io/cluster: ${HostedControlPlane Namespace}`.

For the `ControllerAvailabilityPolicy` option, use `HighlyAvailable`. When you use that option, you can schedule pods for each deployment within a hosted cluster across different failure domains by setting `topology.kubernetes.io/zone` as the topology key.

> //lahinson - sept. 2023 - If we want to explain zone versus hostname, I'll need info for that.

Reviewer 2: @lahinson labeling by hostname is useful in on-prem deployments (e.g., agent/kubevirt) as this deployments has not concept of zones, unlike AWS for example. 

@csrwng @enxebre can you help clarify this as well?

Reviewer: The zone label is required for us to successfully deploy HighlyAvailable clusters. It is up to the customer to decide what zone means. It could be a location or rack name, or at the very least, the hostname of the node where it is set. Without the zone label we cannot successfully deploy HighlyAvailable control planes.

> For the `ControllerAvailabilityPolicy` option, use `HighlyAvailable`. When you use that option, you can schedule pods for each deployment within a hosted cluster across different failure domains by setting `topology.kubernetes.io/zone` as the topology key.

Reviewer 2: We need to call out that `HighleyAavilable` is the default and what is the `HCP` CLI deploys. Non Highley Available control-planes are not supported beyond testing.

Author: I updated the text as follows:
>For the `ControllerAvailabilityPolicy` option, use `HighlyAvailable`, which is the default value that the hosted control planes command line interface (`hcp`) deploys. When you use that option, you can schedule pods for each deployment within a hosted cluster across different failure domains by setting `topology.kubernetes.io/zone` as the topology key. Control planes that are not highly available are not supported.

[#hosted-cluster-workload-distributing]
= Distributing hosted cluster workloads

Before you get started with hosted control planes for {ocp-short}, you must properly label nodes so that the pods of hosted clusters can be scheduled into infrastructure nodes. Node labeling is also important for the following reasons:

* To ensure high availability and proper workload deployment. For example, you can set the `node-role.kubernetes.io/infra` label to avoid having the control-plane workload count toward your {ocp-short} subscription.

> * To ensure that control plane workloads are configured at the "Everything shared" multi-tenancy distribution level, where control planes for hosted clusters can run on any node that is designated for control planes

Reviewer 2: I would remove this completely. We haven't defined "Everything Shared" anywhere, and we are not explaining it enough (makes sense not to because its only relevant for request serving components in a different RH context).

Author: @zanetworker Removed ✅

[#hosted-cluster-workload-distributing]
= Distributing hosted cluster workloads

Before you get started with hosted control planes for {ocp-short}, you must properly label nodes so that the pods of hosted clusters can be scheduled into infrastructure nodes. Node labeling is also important for the following reasons:

* To ensure high availability and proper workload deployment. For example, you can set the `node-role.kubernetes.io/infra` label to avoid having the control-plane workload count toward your {ocp-short} subscription.
//lahinson - sept. 2023 - commenting out the following lines until those levels are supported for self-managed hypershift
//* To ensure that control plane workloads are configured at one of the following multi-tenancy distribution levels:
//** Everything shared: Control planes for hosted clusters can run on any node that is designated for control planes.
//** Request serving isolation: Serving pods are requested in their own dedicated nodes.
//** Nothing shared: Every control plane has its own dedicated nodes. 

> * To ensure that control plane workloads are separate from other workloads in the management cluster

Reviewer 4: missing period at the end. I believe this bullet point on line 12 is _not_ commented out, therefore, please include a period at the end.

*Important:* Do not use the management cluster for your workload. Workloads must not run on nodes where control planes run.

[#hosted-cluster-labels-taints-overview]
== Labels and taints for management cluster nodes

> As a management cluster administrator, you can use the following labels and taints in management cluster nodes to schedule the control plane workload:

Reviewer 4: Optional:

For minimalism purposes, you can remove `you can` especially since you already specified who the user is. Also, can we say, `your control plan workload` or does the user not own this workload? So it'd be: 

As a management cluster administrator, use the following labels and taints in management cluster nodes to schedule your control plane workload:

Author: I like your revision! I'm not sure if it's entirely correct to refer to the control plane workload as "your control plane workload", although I like your intention of being more engaging with the reader. I think changing it to "a control plane workload" might work.

This way, hosted control planes for each tenant are eligible infrastructure node workloads, and you do not need to entitle the underlying {ocp-short} nodes.

> //lahinson - sept. 2023 - in the line above, should we replace "tenant" with something else? Maybe "hosted cluster"?

Reviewer 2: Makes sense :)
